### ISSUE TITLE
Feature: collapsible_heading_for_group + all_collapsed=TRUE

### ISSUE THREAD

**Issue by sda030** [NIFU-NO/saros.base#72]
https://quarto.org/docs/authoring/callouts.html
Use collapse="true" as default
Need an argument in draft_report for what to collapse (each graph, each section, etc.

EDIT: Wait until implemented for headings

**Comment by sda030** [comment]
## Feasibility & Analysis

### What This Involves
Adding collapsible sections (Quarto callouts with `collapse="true"`) to group report content, with an `all_collapsed=TRUE` option. The issue notes waiting until heading-level implementation is done.

### Feasibility
**Medium** — Quarto's [collapsible callout](https://quarto.org/docs/authoring/callouts.html) syntax is straightforward. The main complexity is integrating it into the chunk/heading generation pipeline.

The insertion points are:
- `insert_section_heading_line.R` — wraps headings, could add callout opening syntax
- `insert_chunk.R` — generates chunk content, would need closing syntax
- `gen_qmd_structure.R` — recursive structure generation, needs to track open/close state

### Risks
- Nested collapsibles (section > subsection > chunk) can produce confusing UX
- Quarto callout syntax has limitations — it's div-based (`::: {.callout-note collapse="true"}`), which may not nest cleanly with heading levels
- Interactive elements inside collapsed sections (e.g., plotly) may not render correctly when initially collapsed

### Benefits
- Long reports become much more navigable
- Users can focus on sections of interest
- Reduces visual overload for reports with many variables

### What Needs Changing
1. New `draft_report()` parameters: `collapsible_heading_for_group` (which grouping levels to collapse) and `all_collapsed` (default state)
2. Modify `insert_section_heading_line.R` to optionally wrap output in Quarto callout div syntax
3. Ensure `gen_qmd_structure.R` properly closes div blocks at each level
4. Add chunk template variants that support collapsible wrappers

### Decisions to Make
- Collapse at which level? (chapter, section, element, individual chart?)
- Use Quarto callouts (`.callout-note`) or raw HTML `<details>` elements?
- Should the collapse title show a summary (e.g., N respondents, key finding)?
- How does this interact with tabsets (#50)?

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by sda030** [comment]
Re-analysis against `main` @ `1a3f207`. Everything below is either a file/line reference or the output of a command I ran (Quarto 1.9.14, pandoc 3.6.3, R 4.6.0). This supersedes the 2026-03-21 comment, which was posted without verification; where it was right, that is said plainly.

## 1. The prerequisite: what "wait until implemented for headings" refers to

**It is not the `*_heading_for_group` family, and the blocker is upstream, not here.**

The edit history dates the EDIT precisely:

```
gh api graphql -f query='{repository(owner:"NIFU-NO",name:"saros.base"){issue(number:72){userContentEdits(first:10){nodes{editedAt diff}}}}}'
  createdAt 2024-08-01T21:18:17Z (issue opened)
  editedAt 2024-08-01T21:29:00Z  (body without the EDIT line)
  editedAt 2025-10-26T19:46:41Z  (adds "EDIT: Wait until implemented for headings")
```

The `*_heading_for_group` machinery landed in `bdd82af` (author + commit date `2024-08-01 22:03:22 +0200` = 20:03 UTC), closing #10 and #17 (both `closedAt 2024-08-01T20:09:03Z`) — **one hour before this issue was even opened**, and 14 months before the EDIT. So the EDIT cannot be waiting on `ignore_/replace_/prefix_/suffix_heading_for_group`; those already existed on both dates.

The reading that survives is the literal one: wait until *collapse* is implemented **for headings** — i.e. by Quarto, so that a section can be made collapsible by its heading rather than by wrapping it in a div. That is still not implemented: [quarto-cli discussion #6038](https://github.com/quarto-dev/quarto-cli/discussions/6038) (`Title {.collapse}`) was closed as a duplicate of #5523 ("Allow qmd sections / subsections to be enclosed in `<details>` tags"), with no implementation. Quarto 1.9.14 still offers `collapse` only on `.callout-*` divs.

**Verdict: the prerequisite is *not* satisfied if read as "wait for Quarto".** But it does not have to block the issue, because the div-wrapping workaround is measurable and one of the two candidate mechanisms turns out to be clean (§5). Recommend the maintainer confirm the reading; it is the single decision that determines whether this is actionable. If the intent was in fact "wait for the heading arguments", that is satisfied and has been since day one.

## 2. Verification of the 2026-03-21 analysis

| Claim | Verdict | Evidence |
|---|---|---|
| Insertion point `insert_section_heading_line.R` | **Partly right** | The heading text is built there (`R/insert_section_heading_line.R:61-70`), but the *decision whether a heading is emitted at all* lives in `R/attach_new_output_to_output.R:9-15`. An opening fence must be gated by that same predicate, so the real edit is in `attach_new_output_to_output()` / `gen_qmd_node()`, not in `insert_section_heading_line()`. |
| Insertion point `insert_chunk.R` ("would need closing syntax") | **Wrong** | `insert_chunk()` (`R/insert_chunk.R:24-70`) only glues one `.template` for the *deepest* level. A section's closing fence must come after its entire **subtree**, which `insert_chunk()` never sees. Putting the close there would close a level-2 wrapper before its level-3 children were emitted. |
| `gen_qmd_structure.R` "needs to track open/close state" | **Right conclusion, obsolete mechanism** | The file was rewritten by #19/#238 (shared `gen_qmd_node()` + two engines) and #239/#240 (removed `new_out` threading). No global open/close state is needed or wanted: the pairing is already structural (§3). |
| Callout syntax "is div-based, may not nest cleanly with heading levels" | **Right, and far worse than stated** | Measured: a heading as the **first element** inside a callout is consumed as the callout title and its `{#sec-}` anchor is destroyed — `quarto render` emits `WARNING ... Unable to resolve crossref @sec-a` and the reference renders as `?@sec-a`. Any heading nested *deeper* inside a callout survives as a real `<section>` with a working xref but is **silently dropped from the TOC** (verified against a callout-free control: the same `### Sub A1` appears in the TOC without the callout, and does not with it). |
| Interactive elements "may not render correctly when initially collapsed" | **Not reproduced; unsupported as stated** | A `::: {#fig-...}` float and a `::: {.panel-tabset}` both render inside a `collapse="true"` callout, with `@fig-one` resolving to "Figure 1". There *is* a known ggiraph/htmlwidget sizing problem in hidden containers, but nothing in this repo or my tests demonstrates it, and the original comment offered no evidence. Treat as an open risk to test, not a finding. |
| New params `collapsible_heading_for_group` + `all_collapsed` | **Right** | Fits the existing `*_heading_for_group` convention (`R/draft_report.R:265-279`). |
| Effort: Medium | **Agree** | See §8. |

## 3. The real insertion points

Emitted structure today (from `tests/testthat/_snaps/qmd_snapshots.md`): the chapter `#` is written directly by `R/gen_qmd_chapters.R:129`; every deeper heading comes from the traversal; `.chapter_number`/`chapter` are in `ignore_heading_for_group` by default (`R/draft_report.R:265-272`, the #207 fix). Default `organize_by` (`R/refine_chapter_overview.R:256-261`) makes the **deepest** level `.template_name`, whose heading is also ignored — so the sensible collapse levels are the middle ones (`.variable_label_prefix_dep` → `##`, `.variable_name_indep` → `###`).

**Open fence** — `R/gen_qmd_structure.R:63-83`, immediately after `insert_section_heading_line()` and inside the same call that folds the heading in, `attach_new_output_to_output()` (`R/attach_new_output_to_output.R:9-15`). It must be emitted under *exactly* the condition that emits the heading.

**Close fence** — the point where a completed child's return value reaches its parent, which exists in both engines and is one-to-one with `gen_qmd_node()`:

* recursion: `R/gen_qmd_structure.R:153-154` (`added <- walk(...)`; `combine_child_output()`)
* loop: `R/gen_qmd_structure.R:207-211` (the `have_pending` fold)

The pairing is **total**: even at the deepest level the child frame is `beyond` and returns `character()` (`R/gen_qmd_structure.R:197-205`), and the parent still folds it. So every `gen_qmd_node()` call has exactly one matching fold, in both engines. That is the structural guarantee this feature needs, and it already exists.

**The one asymmetry that matters.** At the recursion's fold point `node` is still in scope, so `node$close_fence` is available directly. At the loop's fold point it is **not** — the frame was re-pushed at `:238-241` and the loop restarted. The loop engine must stash the close on the parent frame (a new `frame$pending_close` field) and consume it at `:207-211`. This is the only place the two engines' code would differ, and therefore the only place they can drift. `tests/testthat/test-qmd_engines.R` is exactly the guard for it.

**Design rule:** `gen_qmd_node()` must be the single place that decides both halves, returning the matching close in its result list (`R/gen_qmd_structure.R:85-89`). If the open is decided in `attach_new_output_to_output()` and the close is decided independently at the fold point, they are two expressions evaluated at different times over different data — that is the bug.

Note also that `combine_child_output()` (`:94-98`) calls `stri_remove_empty_na(added)`, so a close passed through it as part of `added` survives, but a close for a subtree that emitted nothing would need to be appended separately rather than merged.

## 4. What guarantees the fences balance today: **nothing** — and they already don't

The traversal emits no fences at all. Every `:::` in a generated `.qmd` comes verbatim from `.template` via `glue_data()` at `R/insert_chunk.R:67`. There is no balance check anywhere.

Two shipped default templates are **already unbalanced**:

```
# scan of the template store
variant 1 row 4 name=int_table_html indep=NA opens=0 closes=1   # R/zzz.R:159-178
variant 5 row 4 name=int_plot_html  indep=NA opens=0 closes=1   # R/zzz.R:1088-1105
```

and it reaches users. Generating a report over the numeric columns of `ex_survey`:

```r
ov <- data.frame(chapter = "Num", dep = "c_1, c_2", indep = "")
# ... refine_chapter_overview() -> draft_report()
1_Num.qmd  opens=0  closes=1
```

Rendering that stray `:::` gives `The following string was found in the document: :::  This usually indicates a problem with a fenced div in the document.` plus a literal `<p>:::</p>` in the body. **This is a live pre-existing bug and should be split into its own issue** — but it is also the proof that the property test proposed in §7 is not hypothetical: it fails on `main` today.

**What #239 implies for balance.** `R/attach_new_output_to_output.R:9-11` emits the heading iff the node produced a chunk **or** `level < length(grouping_structure)`. So:

* At a **deepest**-level node with no chunk, *nothing* is emitted — not even the heading (`tests/testthat/test-qmd_empty_section.R:103-114`). An open fence tied to the heading is therefore conditional, while a close emitted at the parent's fold point is unconditional. **That is the exact shape of the hazard**, and it is why the close must be returned by `gen_qmd_node()` rather than recomputed.
* At an **intermediate** node the heading is emitted unconditionally, even when the whole subtree is empty. I confirmed this in both engines with a 3-level fixture whose middle branch matches no rows — both emit a bare `## {#sec-m2-3bd770}` (an empty heading; a separate cosmetic defect worth its own issue). Balance is preserved there, but you would get an empty collapsed box.

**The opening half is already user-reachable.** `suffix_heading_for_group` is inserted directly after the heading line (`R/insert_section_heading_line.R:61-70`), so a user can do this today:

```r
draft_report(..., suffix_heading_for_group =
  c(".variable_label_prefix_dep" = "\n::: {.callout-note collapse=\"true\"}\n"))
# 1_Bakgrunn.qmd -> opens=3 closes=2
```

An unbalanced open, silently. Framed correctly, **this feature is "supply the matching close"**, not "add fences".

## 5. Callouts vs `<details>` — measured, then a recommendation

| | `::: {.callout-note collapse="true"}` | `<details>`/`<summary>` in a `` ```{=html} `` block |
|---|---|---|
| Heading as **first** element inside | consumed as the callout title; `{#sec-}` destroyed; `@sec-a` → `?@sec-a` + `Unable to resolve crossref` | n/a (heading stays outside) |
| Heading **nested** inside | real `<section>`, xref works, **absent from the TOC** | real `<section>`, xref works, **present in the TOC** |
| `::: {#fig-}` float inside | works (`@fig-one` → Figure 1) | works |
| `::: {.panel-tabset}` inside | renders | renders |
| docx | renders expanded (`collapse` ignored); content preserved | wrapper dropped cleanly; content preserved |
| typst | renders as `#callout(...)`, expanded | wrapper dropped cleanly |
| pptx | renders as a styled block on the slide | wrapper dropped cleanly |
| revealjs | supported "without collapse option" (Quarto docs) | dropped |

Two caveats on `<details>`. Written as bare markdown lines, the `<summary>` text **leaks as a plain paragraph** into docx/typst/pptx (verified: `Details wrapper` appears as a bare paragraph in `t4.typ` and in the docx text stream). Wrapping the tags in `` ```{=html} `` fences fixes that completely — in that form typst and docx contain no trace of the wrapper, and the HTML still nests correctly (`</section></details></section>`, well-formed). Second, a collapsed callout with no `title=` still gets a working toggle, but with a blank title bar.

**Recommendation: `<details>`/`<summary>` inside `` ```{=html} `` raw blocks.**

1. The TOC is the primary navigation device in a saros report. Callouts silently remove every nested heading from it; `<details>` does not. That alone is decisive.
2. #213 exists to make `{#sec-}` anchors stable and cross-referenceable. A callout-based design destroys the anchor of whichever heading it wraps first — silently undoing #213 for exactly the levels the user asked to collapse.
3. Callouts impose note/tip iconography and colouring on every ordinary section, which is semantically wrong for "this is just a section".
4. Non-HTML degradation is *cleaner* for the raw-block `<details>` (wrapper vanishes entirely) than for callouts (an expanded, styled, unremovable box in every docx/typst/pptx section).

Cost of that choice, stated honestly: `<details>` is not themed, gets no Quarto styling, is invisible to revealjs, and needs a `<summary>` label the caller must supply or the generator must synthesise. If the maintainer prefers the Quarto-native look and accepts the TOC loss, callouts remain viable — but then the heading must stay **outside** the fence and the callout needs an explicit `title=`, never a heading as its first child.

## 6. Tabsets (#50) and output formats

* Variant 2 (`R/zzz.R:265-558`) does not put `::: {.panel-tabset}` in the `.qmd`; it emits `#| panel: tabset` inside an `output: asis` chunk that generates `##### <name>` headings via `knitr::knit_child()` at **render** time. So the tabset div is created inside whatever fenced div is open at that point. With a callout that means the runtime `#####` headings land inside it — harmless for the TOC (they were never in it) but it makes the callout's own title resolution order-dependent if no `title=` is given. With `<details>` in raw blocks there is no interaction at all: knitr's chunk detection is line-based and unaffected by raw HTML.
* Variant 3 (`R/zzz.R:562-798`) puts a literal `::: {.panel-tabset}` in the `.qmd`. Verified to render inside a collapsed callout.
* Formats: `collapse` is HTML-only in every case. Under docx/typst/pptx a callout becomes a permanently-expanded styled block; a raw-block `<details>` disappears and the content renders as ordinary flow. Neither errors.

## 7. Testability

Fence balance is a **plain textual property of the emitted `.qmd`** and should be asserted as one, not left to a snapshot:

```r
opens  <- grep("^:{3,}[ ]*[{]", lines)   # ::: {...}
closes <- grep("^:{3,}[ ]*$",   lines)   # bare :::
expect_identical(length(opens), length(closes))
```

A stack walk over the same two patterns additionally proves no close precedes its open. This is strictly stronger than a snapshot: a snapshot pins one fixture's bytes and can be `snapshot_accept()`-ed past, whereas the property holds for every fixture and every engine and states the intent. It is the same discipline `tests/testthat/test-qmd_snapshots.R:221-223` already applies to heading whitespace ("Stated as a property too, so the intent survives a careless snapshot accept").

Concretely:

* new `tests/testthat/test-qmd_fences.R` — balance + ordering over every generated `.qmd`, run under **both** `qmd_engine` values, with fixtures that include an empty deepest-level section (reusing the `test-qmd_empty_section.R` phantom-path construction) and an empty intermediate branch;
* a template-store scan in `tests/testthat/test-chunk_templates.R` — same source-level rationale as the #214 guard: it covers every template in every variant, not just those a fixture happens to instantiate. **This will fail immediately on the two templates in §4**, which is the point;
* extend `test-qmd_engines.R` with a `collapsible_heading_for_group` case — the loop engine's `frame$pending_close` is the drift risk;
* re-record `test-qmd_snapshots.R` with one collapsed fixture, and verify each new test fails against the pre-change code before accepting it.

## 8. Verdict

* **Effort: Medium** — the traversal change is genuinely small (one field returned from `gen_qmd_node()`, one consumed at each engine's fold point, one field on the loop frame), but the design fork, the format matrix and the fence-balance test surface put it at roughly a day. **Agrees with the existing `Effort: [2] Medium` label.** It would be Small if the mechanism were already decided and the `<details>` route taken.
* **Files touched:** `R/gen_qmd_structure.R` (return + consume the close in both engines), `R/attach_new_output_to_output.R` (emit the open under the heading predicate), `R/insert_section_heading_line.R` (label for `<summary>`/`title=`), `R/gen_qmd_chapters.R` + `R/draft_report.R` + `R/validate_draft_report_args.R` (thread and validate), `man/` + `NAMESPACE` (CI fails on roxygen drift, #219), `NEWS.md`, `tests/testthat/test-qmd_fences.R` (new), `test-qmd_engines.R`, `test-qmd_snapshots.R`, `test-chunk_templates.R`. Not `R/insert_chunk.R`.
* **New API surface:** `collapsible_heading_for_group` (character vector of grouping columns, `NULL` default, mirroring `ignore_heading_for_group`); `all_collapsed` (scalar logical); probably `collapsible_style = c("details", "callout")` so the fork is a user choice rather than a permanent bet; optionally a `collapsible_label_for_group` named vector for the `<summary>`/`title=` text. Internal: one `collapsible_fences()` helper returning the open/close pair.
* **Test strategy:** property assertion on emitted `.qmd` text (opens == closes, no close before its open), run under both engines and over fixtures with empty deepest and empty intermediate sections; plus a template-store scan; plus snapshots. **Yes — the property is directly checkable in emitted `.qmd` text**, with two `grep()`s.
* **Unknowns requiring a maintainer decision:**
  - Which reading of "wait until implemented for headings" was intended (§1). Decides whether this is actionable now.
  - Callouts vs `<details>` (§5) — accepting the TOC loss is the crux.
  - Whether `collapsible_heading_for_group` should be *ignored* for a level that is also in `ignore_heading_for_group` (a wrapper with no heading), or whether that is a legitimate "collapse an unlabelled block" case.
  - What the `<summary>`/`title=` text should be: the heading text (duplicated immediately under the real heading), a fixed string, or caller-supplied.
  - Whether an intermediate node with an empty subtree should still open a wrapper (currently it would emit an empty collapsed box, since its heading is emitted unconditionally).
  - Whether to fix the two unbalanced default templates (§4) here or in a separate issue — I recommend separate.
* **Risk of silent wrongness: high, and asymmetric — this is the main argument for the property test.**
  - An **unmatched close** is loud-ish: Quarto warns `This usually indicates a problem with a fenced div` and a literal `:::` appears in the body. Ugly but visible. (This is the failure mode already shipping.)
  - An **unmatched open** is the dangerous one. Pandoc emits only `[WARNING] Div at line 5 column 1 unclosed at line 18 column 1, closing implicitly.` — buried in render logs — and then swallows **the entire remainder of the chapter** into the collapsed wrapper. I verified this: sections after the unclosed fence render inside the collapsed callout and vanish from the TOC. The page looks entirely plausible; the content is simply hidden. That is the same failure signature as #239 — correct-looking output, wrong content — and it is why the close must be produced by the same expression that produces the open, and why the balance property belongs in the test suite rather than in a reviewer's eye.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
